core/src/main/scala/kafka/server/FetchSession.scala

```diff
@@ -273,9 +273,9 @@ trait FetchContext extends Logging {
   def getFetchOffset(part: TopicPartition): Option[Long]
 
   /**
-    * Apply a function to each partition in the fetch request.
+    * Get all of the partitions in the fetch request.
     */
-  def foreachPartition(fun: (TopicPartition, FetchRequest.PartitionData) => Unit): Unit
+  def partitions: Vector[(TopicPartition, FetchRequest.PartitionData)]
```

Comment:
Why do you want to materialize all the partitions?  This is less efficient since it means another copy in memory.

Reply:
I'll change it return an `Iterator`. Different implementation can decide what is efficient in their case. I wanted to avoid executing unknown code in this [pattern](https://github.com/apache/kafka/pull/6261/files#diff-30e592741598aded2913e8834ea0e372R406). The issue here is that at best we have lock contention and at worst a deadlock.

What do you think?